### DISCUSSION TITLE
Announcement: pXML becomes PDML

### DISCUSSION THREAD

**Discussion by pml-lang** [pml-lang/pml-companion#62] [Show and tell]
_practicalXML (pXML)_ (the syntax used in PML) has been renamed to _Practical Data and Markup Language (PDML)_.
The documentation has been improved, and everything is now published on the new [PDML website](https://pdml-lang.github.io/index.html).

Some interesting links:
- [Introduction to PDML](https://pdml-lang.github.io/docs/introduction/index.html)
- [Specification](https://pdml-lang.github.io/docs/basic-specification/index.html)
- Grammar defined with [EBNF](https://pdml-lang.github.io/docs/ebnf/index.html) and [railroad diagrams](https://pdml-lang.github.io/docs/railroad/index.html)
- [Examples](https://pdml-lang.github.io/docs/examples/index.html)
- [Github repository](https://github.com/pdml-lang) (parser source code, website files, and some documents written in PML)

I also plan to write a document covering all aspects of PDML _extensions_ (attributes, extension nodes, raw text blocks, types, etc.)

Feedback is very welcome.

**Comment by tajmone** [comment]
Very cool! I've read the docs and specs, and I grasp how simple yet powerful it is compared to XML, JSON, YAML and company.&trade.

And it's very easy to parse (unlike YAML), which equates to simple to implement editor syntaxes.

Serialization and deserialization to other configuration formats also seems very easy to achieve — and, most important, the process is never lossy on the PDML side, for it's capable of fully encompassing its counterparts. I think it could easily become the new adopted standard in large projects configurations, especially where time and space optimization is crucial (e.g. videogames that rely on external configs to mod their internal configuration).

**Comment by Zegnat** [comment]
I just stumbled upon PML and PDML and think this is great. My hope would be for PDML to really be "the syntax used in PML" so a single parser could be created. I guess this requires attributes to be specced?

Do you envision PML to be just a set of node definitions for a PDML document, or do you always imagine PML will lead and PDML will follow?

**Reply by pml-lang** [reply]
> I just stumbled upon PML and PDML and think this is great.

Thank you very much.

> My hope would be for PDML to really be "the syntax used in PML" so a single parser could be created.

This is the case already. PML is based on PDML. PML syntax _is_ PDML syntax. And the PML Converter uses the PDML parser. There is no specific PML parser (although that was the case in the first versions of PML, before the creation of PDML).

> I guess this requires attributes to be specced?

Support for attributes is implemented already in the PDML parser, so they can be used in PML too. However, attributes are not part of _basic_ PDML. Attributes belong to a set of optional PDML _extensions_. PDML extensions are currently not yet documented on PDML's website, but I am working on this. For now you can have a look at chapter [Extensions](https://pdml-lang.github.io/docs/introduction/index.html#extensions) in [Introduction to PDML](https://pdml-lang.github.io/docs/introduction/index.html).

> Do you envision PML to be just a set of node definitions for a PDML document

Yes, you can look at it like this. PDML is an extensible format without rules about which nodes/attributes you can use. On the other hand,  PML defines a set of valid nodes/attributes, and each node has specific semantics and behavior.

> or do you always imagine PML will lead and PDML will follow?

PML depends on PDML. And PDML does not depend on PML.

Thanks for your interest in PDML/PML.

**Reply by Zegnat** [reply]
Oh, I totally missed that Extensions page, as I only looked through the PDML Specification. Which I now realise is titled ”Basic PDML Specification“. Thank you for pointing that out!

Just from a quick read through, is it true that a _basic_ PDML parser is expected to not parse PDML documents that include some of those extensions? This feels like it could lead to very split communities and incompatible parsers. From my reading, the only extension that would degrade gracefully and not have a parser give up are attributes.

| PDML | Basic Parser | Parser with extensions |
| --- | --- | --- |
| `[a b=c d]` | Node `a` with single child text node “b=c d” | Node `a` that has an attribute named `b` with the value “c” and a single child text node with content “d”. |
| `[- comment -]` | Parser error, node name `-` not allowed. | Comment node (no parsing defined for contents) |
| `[!set a=b]` | Parser error, node name `!set` not allowed. | Parameter `a` is set to value “b” for later use |
| `line \u0041 \U0001F600` | Parser error, unescaped `\` not allowed. | Text node with content “line A 😃”|

(What applies to `[!set]` applies to any nodes starting with `!`, of course.)

The `[code]` example will mostly work in basic parsers from my reading. You just end up with a tonne more whitespace in your text nodes. Which I think is fine for basic parsers to do, and then to be cleaned up by parsers that have a code extension. But it will only work as long as you do not have any disallowed strings in there (only `\`, I think?) and no illegal nodes (`[Hello]` is fine and will get parsed as an empty node, while JS code like `["a", "b"]` will result in a parser error). The idea of `"""` seems great to me. I guess that would be PDML’s version of XML’s CDATA? An instructor that says not to parse the contents for PDML syntax.

It sounds like it would be good to reserve certain node names specifically for extensions. Allowing node names to start with `-` and `!` would be enough in most cases for basic parsers to not hard fail. Having parsers give up half way through when a user believes they have written valid PDML seems like a worse experience to me.

I think the idea of a simple base spec and offloading lots of features to optional extensions is great. But the way this looks now is that there are two completely separate grammars/syntaxes here. The <small>[EA]?</small>BNF of Basic PDML cannot currently validate a PDML document with most of those extensions.

**Reply by pml-lang** [reply]
> is it true that a basic PDML parser is expected to not parse PDML documents that include some of those extensions?

Yes. All syntax extensions should _fail-fast_ if they are parsed by a _basic_ PDML parser. This is to avoid that a basic PDML parser silently ignores incompatibilities and produces a wrong result. For example, all extension nodes start with `[!`. Thus, a basic parser immediately raises an error, because the `!` is an illegal character in basic PDML. This eliminates the risk of silently producing a wrong result, and ensures that the user gets immediatley noticed about the problem.

There are corner cases, however, as well illustrated in your comment.

For example, this code:
```
[foo (n=v)]
```
... is ambiguous.
A basic parser parses this as 'node `foo` with text `(n=v)`'
An extended parser supporting attributes parses this as 'node `foo` with attribute `n` set to `v`'

If `(n=v)` is meant to be text you can escape the `(` to eliminate the ambiguity:
```
[foo \(n=v)]
```

However, if `(n=v)` is meant to be an attribute assignment then a basic parser will parse it as text, and there is currently no way to avoid this.

I was thinking already about changing the syntax for attributes, and use an extension node with name `a` for attributes. That would mean code like this:
```
[foo (n1=v1 n2=v2)]
```
... would change to:
```
[foo [!a n1=v1 n2=v2]]
```
This would eliminate the above ambiguity, but it would also make the syntax for attributes more verbose. I'm not convinced that this would be a good idea. Attributes are used often, and their syntax should stay simple. A very rare corner case like the above one does (IMHO) not justify the increased verbosity.

Note: Escaping a `(` with `\(` is implemented already in the parser, but it is not mentioned yet in the documentation. I'll fix this. You can try it out in PML (which uses extended PDML) with code like this:
```
[div \(n=v)]
```
Alternatively you can write:
```
[div() (n=v)]
```
... to make it clear that there are no attributes and `(n=v)` is text.

> The [code] example will mostly work in basic parsers

This is another good example of possible corner cases.

The lesson is clear: PDML documents using PDML extensions should never be parsed with a _basic_ PDML parser. If this is done accidentally, then a syntax error will hopefully be raised by the parser. However, in some rare corner cases, a wrong result might be produced.

> The idea of """ seems great to me. I guess that would be PDML’s version of XML’s CDATA?

Yes. """ is more user-friendly and more powerful than a CDATA section. The rules for 'raw text' extension nodes will soon be documented on PDML's website.

> reserve certain node names specifically for extensions

All extension nodes must start with `[!`. Hence there will never be a name clash of an extension node and a data node.

> The [EA]?BNF of Basic PDML cannot currently validate a PDML document with most of those extensions.

Yes, and that's deliberate, because it supports the important [fail-fast](https://medium.com/@christian.ppl/introduction-to-the-fail-fast-principle-in-software-development-865ccab28979) principle.

**Reply by tajmone** [reply]
> @Zegnat — Just from a quick read through, is it true that a basic PDML parser is expected to not parse PDML documents that include some of those extensions? This feels like it could lead to very split communities and incompatible parsers.

Regarding the different ways whitespace can be handled (i.e. ignored vs parsed as sub-nodes), this is a question which came up in various discussions and (IIRC) the general idea is that probably the PDML standard will branch into sub-standards based on domain specific applications.

The whitespace handling problem is a good example of how different application domains might split regarding how basic parsing rules are applied. PDML being an open standard, it will probably embrace changes based on real case usage, and therefore come up with sub-specs that are domain specific.

There's always going to be an inevitable tension between the specs and how specific parsers are implemented, with the intrinsic risk of "very split communities and incompatible parsers". What the PDML standard can do, through communication with its users, is to try to embody the community needs into its official specs, thus mitigating incompatibility issues.

But as far as proliferation of different communities, I think that's a desirable side-effect, which is why there's a "PDML Basic Spec" to start with — meaning it doesn't end there, and more specs are to come.

Definitely, the idea of a «branching specs» has come out in PDML discussion various times, and it seems a rather natural phenomenon in this case since different application domains face different needs.

**Reply by pml-lang** [reply]
> What the PDML standard can do, through communication with its users, is to try to embody the community needs into its official specs, thus mitigating incompatibility issues.

Well said!

I will soon (probably this month) publish a first version of "PDML Extensions", a document that will evolve and improve over time, based on real world experiences and feedback from the community.
The challenge will be to provide a specification that ensures compatibility as far as possible, while at the same time supporting domain-specific customization in a well-defined way.
